runtimes/protocol/notification.ts

```diff
@@ -0,0 +1,69 @@
+import { MessageType, ProtocolNotificationType } from './lsp'
+
+export interface NotificationContent {
+    readonly text: string
+    readonly title?: string
+}
+
+export namespace FollowupNotificationActionType {
+    export const Acknowledge = 'Acknowledge'
+}
+
+export type FollowupNotificationActionType = typeof FollowupNotificationActionType.Acknowledge
+
+export namespace NotificationActionType {
+    export const Url = 'Url'
+    export const Marketplace = 'Marketplace'
+}
+
+export type NotificationActionType =
+    | typeof NotificationActionType.Url
+    | typeof NotificationActionType.Marketplace
+    | typeof FollowupNotificationActionType.Acknowledge
+
+export interface NotificationAction {
+    readonly text: string
+    readonly type: NotificationActionType
+}
+
+export interface UrlAction extends NotificationAction {
+    readonly type: typeof NotificationActionType.Url
+    readonly url: string
+}
+
+export interface MarketplaceAction extends NotificationAction {
+    readonly type: typeof NotificationActionType.Marketplace
+}
+
+export interface AcknowledgeRequestAction extends NotificationAction {
+    readonly type: typeof FollowupNotificationActionType.Acknowledge
+}
```

Comment:
How are these to be used/why define them?

Reply:
Exactly how they are used now - to pass the type of action to the client, so that client can decide what to do with it.

Reply:
Ah, I see now.  They would be passed in NotifactionParams.actions as concrete NotificationAction types.  Didn't catch that on the first pass.